core/src/main/java/com/epam/esm/dao/impl/GiftDaoImpl.java

```diff
@@ -0,0 +1,282 @@
+
+        for (GiftCertificateEntity giftCertificateEntity : query){
+            Set<TagEntity> tags = getTagsByGiftId(giftCertificateEntity.getId());
+            giftCertificateEntity.setTags(tags);
+        }
+        return query;
+    }
+
+    @Override
+    public GiftCertificateEntity createGift(GiftCertificateEntity entity){
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        Timestamp currentTimestamp = new Timestamp(System.currentTimeMillis());
+        jdbcTemplate.update(
+                connection -> {
+                    PreparedStatement ps =
+                            connection.prepareStatement(INSERT_GIFT_QUERY, new String[] {COLUMN_ID});
+                    ps.setString(1, entity.getName());
+                    ps.setString(2, entity.getDescription());
+                    ps.setInt(3, entity.getPrice());
+                    ps.setInt(4, entity.getDuration());
+                    ps.setTimestamp(5, currentTimestamp);
+                    ps.setTimestamp(6, currentTimestamp);
+
+                    return ps;
+                },
+                keyHolder);
+
+
+        Long giftId = (Long) keyHolder.getKey();
+
+        entity.setId(giftId);
+        entity.setCreateDate(currentTimestamp);
+        entity.setLastUpdateDate(currentTimestamp);
+
+        return entity;
+    }
+
+    @Override
+    public GiftCertificateEntity updateGift(GiftCertificateEntity giftCertificateEntity){
+        Timestamp currentTimestamp = new Timestamp(System.currentTimeMillis());
+        int rows = jdbcTemplate.update(UPDATE_BY_ID_QUERY,
+                giftCertificateEntity.getName(),
+                giftCertificateEntity.getDescription(),
+                giftCertificateEntity.getPrice(),
+                giftCertificateEntity.getDuration(),
+                giftCertificateEntity.getId());
+
+        if (rows == 0){
+            throw new GiftNotFoundException();
+        }
+
+        giftCertificateEntity.setLastUpdateDate(currentTimestamp);
+
+        return giftCertificateEntity;
+    }
+
+    @Override
+    public void deleteGiftById(Long giftId){
+        jdbcTemplate.update(DELETE_GIFT_BY_ID, giftId);
+    }
+
+    private Set<TagEntity> getTagsByGiftId(Long giftId){
+        List<TagEntity> queryTags = jdbcTemplate.query(SELECT_TAGS_BY_GIFT, TAG_ROW_MAPPER, giftId);
+
+        return new HashSet<>(queryTags);
+    }
+
+    private QueryArgModel defineQueryAndArgs(CustomSearchRequest customSearchRequest){
+        String namePrefix = customSearchRequest.getNamePrefix();
+        String descriptionPrefix = customSearchRequest.getDescriptionPrefix();
+        String tagNamePrefix = customSearchRequest.getTagNamePrefix();
+        String sortField = customSearchRequest.getSortField();
+        String sortMethod = customSearchRequest.getSortMethod();
+        Object[] args = new Object[0];
+
+
+        String namePrefixWithWildCard = ZERO_OR_MORE_ELEMENTS_WILDCARD + namePrefix + ZERO_OR_MORE_ELEMENTS_WILDCARD;
```

Comment:
when I was asking to build sql script I mean the following:
you have the base script `"SELECT gc.id, gc.name, gc.description, gc.price, gc.duration, gc.create_date, gc.last_update_date FROM gift_certificate gc, gift_tags gt, tag t WHERE gc.id = gt.gift_id and gt.tag_id = t.id"`
than based on passed params you append sqs conditions:
```
if (namePrefix != null) {
 builder.append(String.format("and name like %%s%, name)); --- just for example, here can be mistake
}
```
and so on

Reply:
Obviously, you must avoid so many scripts- just one base and additional conditions

Reply:
In the future I may need these requests. It is also convenient to fix errors in queries in one class